### DISCUSSION TITLE
Best practices for functions whose return values vary in type

### DISCUSSION THREAD

**Discussion by covecreekjeremy** [microsoft/pyright#7020] [Q&A]
I'm using pyright on code which parses HTML with lxml, e.g.:

```python
from lxml import etree

def test():
    root = etree.fromstring('<html><body><div></div><div></div></body></html>')
    div_elems = root.xpath('//div')
    print(len(div_elems))
```

On the last line pyright gives:
```
Argument of type "_XPathObject" cannot be assigned to parameter "__obj" of type "Sized" in function "len"
    Type "_XPathObject" cannot be assigned to type "Sized"
        "float" is incompatible with protocol "Sized"
            "__len__" is not present
```

This is a clean codebase and I'd like to get strict typing right from the start. What's the best way to think about a function like `xpath`? Its type stub is right - it can return numbers or lists of strings, depending on how it's called. In practice it only returns one of them and I'd like to benefit from subsequent type checking. If I annotate `div_elems` as `list[str]` then pyright complains about the assignment.

**Comment by erictraut** [comment]
Can you tell me more about `lxml`? Is this the publicly-available package, e.g. [through pypi](https://pypi.org/project/lxml/)? If so, what version do you have installed? I tried installing the latest version (5.1.0), but it doesn't appear to be a typed library. The `etree` submodule is implemented in a native binary, so pyright has no way of understanding the types of the symbols exported by it. Do you have a type stub package for this library?

**Reply by covecreekjeremy** [reply]
Aha yes I'm using lxml 5.1.0 and lxml-stubs 0.5.1, installed via "pip install lxml" and "pip install lxml-stubs" earlier today. I'm happy to edit those type stubs if there's a better way to have pyright interact with a function like xpath()

**Reply by erictraut** [reply]
The type definition of `xpath` in the stubs indicates that it returns a type of `_XPathObject`. This is a union of a bunch of types, some of which will work with `len` and some which will not. To write robust and type-safe code, you'd need to add a check at runtime before calling `len(div_elements)`. The `typing.Sized` class is a runtime-checkable protocol, which means that it works with `isinstance` at runtime. If you're pretty confident that `root.xpath("//div")` will always return an object that is `Sizable`, then you can insert an `assert` into your code to document this assumption (both for human readers and for the type checker).

```python
from typing import Sized
from lxml import etree

def test():
    root = etree.fromstring("<html><body><div></div><div></div></body></html>")
    div_elems = root.xpath("//div")
    assert isinstance(div_elems, Sized)
    print(len(div_elems))
```

**Reply by covecreekjeremy** [reply]
Got it - that's what I was looking for. Thank you!
